micall/core/contig_stitcher.py

```diff
@@ -0,0 +1,631 @@
+    both left to right and right to left, calculating a score that represents a moving
+    average of matches at each position. If characters match at a given position,
+    a score of 1 is added; otherwise, a score of 0 is added. The score is then
+    averaged with the previous scores using a weighted sliding average where the
+    current score has a weight of 1/3 and the accumulated score has a weight of 2/3.
+    This sliding average score is halved and then processed again, but in reverse direction.
+
+    :param left: string representing first sequence
+    :param right: string representing second sequence
+    :return: list representing concordance ratio for each position
+    """
+
+    if len(left) != len(right):
+        raise ValueError("Can only calculate concordance for same sized sequences")
+
+    result: List[Fraction] = [Fraction(0)] * len(left)
+
+    def slide(start, end):
+        scores_sum = Fraction(0)
+        inputs = list(zip(left, right))
+        increment = 1 if start <= end else -1
+
+        for i in range(start, end, increment):
+            (a, b) = inputs[i]
+            current = Fraction(1) if a == b else Fraction(0)
+            scores_sum = (scores_sum * 2 / 3 + current * 1 / 3)
+            result[i] += scores_sum / 2
+
+    # Slide forward, then in reverse, adding the scores at each position.
+    slide(0, len(left))
+    slide(len(left) - 1, -1)
+
+    return result
+
+
+def disambiguate_concordance(concordance: List[float]) -> Iterable[Tuple[float, int]]:
+    for i, x in enumerate(concordance):
+        global_rank = i if i < len(concordance) / 2 else len(concordance) - i - 1
+        yield x, global_rank
+
+
+def concordance_to_cut_points(left_overlap, right_overlap, aligned_left, aligned_right, concordance):
+    """ Determine optimal cut points for stitching based on sequence concordance in the overlap region. """
+
+    concordance_d = list(disambiguate_concordance(concordance))
+    sorted_concordance_indexes = sorted(range(len(concordance)), key=lambda i: concordance_d[i])
+    def remove_dashes(s): return ''.join(c for c in s if c != '-')
+
+    for max_concordance_index in reversed(sorted_concordance_indexes):
+        aligned_left_q_index = len(remove_dashes(aligned_left[:max_concordance_index]))
+        aligned_right_q_index = right_overlap.alignment.query_length - \
+            len(remove_dashes(aligned_right[max_concordance_index:])) + 1
+        aligned_left_r_index = left_overlap.alignment.coordinate_mapping.query_to_ref.left_max(aligned_left_q_index)
+        if aligned_left_r_index is None:
+            aligned_left_r_index = left_overlap.alignment.r_st - 1
+        aligned_right_r_index = right_overlap.alignment.coordinate_mapping.query_to_ref.right_min(aligned_right_q_index)
+        if aligned_right_r_index is None:
+            aligned_right_r_index = right_overlap.alignment.r_ei + 1
+        if aligned_right_r_index > aligned_left_r_index:
```

Comment:
I didn't see any test scenarios that cover this condition. If I force this condition, the tests still pass. Here's what I did to try and find a failing test:
```suggestion
        if True or aligned_right_r_index > aligned_left_r_index:
```

I was looking at this, because I was trying to find any tests of the rules for finding the cut point with the highest concordance. Is that tested?

Reply:
This check (and the whole loop) is for something that I think should never happen in practice, but I cannot prove it. Assuming that it cannot happen, the function seemed too simple for a separate test. I tested the underlying function `calculate_concordance` separately.

Should I still add a test for it?